### ISSUE TITLE
support more distributions

### ISSUE THREAD

**Issue by bobcolner** [getguesstimate/guesstimate-app#43]
Great project.

Please consider supporting more distributions. The types below are useful for 'guesstimate' simulations:
Exponential, Pareto, Generalized extreme value, Truncated Normal, skew normal, t-distribution, Gompertz.

**Comment by davidmanheim** [comment]
Is there a specific syntax for these that would be used? For many "guesstimate" type problems, the key is ensuring that the right distribution types are available - specifically, that we have bounded, unbounded left, and unbounded. There is also a need for a fat tailed distribution. In the application, this probably requires jStat; http://jstat.github.io/

My initial thought for syntax and distributions for how to do this is;

[a,b] is 5%,95% of normal, as currently.
[a,b,c] is a fat tailed distribution, with a, b, as 5% and 95%, and c is 99%.
<a,b> is a truncated normal, with max, min at 5%, 95%.
<x, [a], y> is bounded by x, y, with mean a, so that it could default to, say, a beta scaled to between x and y, and we set alpha=beta.
<x, [a,b]> is log-normal bounded below by x, so it is offset by x, and scaled so that [a,b] are still 5%, 95%.

It might be nice to have a full list of which distributions to use, and every combination of <x,[a,b,c],y>, so that the UI is easy to learn, but that seems more difficult to do statistically; there are probably some annoying numerical computations needed to fit those distributions.

**Comment by OAGr** [comment]
More distributions are definitely a high priority, probably the #1 main feature addition at this points.  There's some tooling around adding them: for those interested, all of this is in the [guesstimator library](https://github.com/getguesstimate/guesstimate-app/tree/master/src/lib/guesstimator).

I didn't even know about jstat, seems very comprehensive.  Does it handle sampling from distributions?

I've been thinking about notation a lot as well.  For fat tailed, maybe the middle one would be the 50%?

I think that the next set of distribution types probably will need quite a bit of planning.  If we go one at a time, we could wind up in a solution that doesn't fit well generally.  Would definitely prefer a challenging codebase to a challenging UI.  

I'm personally not amazing at stats, this is an area I could use some help.

**Comment by davidmanheim** [comment]
@OAGr - I'm potentially available to help (a limited amount,) but I think the first step is the roadmap; what distributions should be included, and what is the syntax? Once that is handled, the code is just churning through that list.

For each distribution, I assume we need code that does the following;
A) Parse the input to pick the distribution desired.
B) Fit the distribution based on the given points. (This means we need to first specify the points that we will get as an input)
C) Sample from the distribution to get our data points.(Of course, the sampling will not be robust in the tails, since sampling well is a trickier thing to do, and we don't want to try to get too fancy, with weighting, etc. - but that could be in a roadmap much further down the line. Having code now that allows it would be nice.)

From there, I'd hope the generated data should be handled in a standardized way; the graphs just take a set of sampled points, etc.

**Comment by bobcolner** [comment]
As for the UI, different distribution types / families have differing amount of parameters -so will need to consider that. 
Another approach would be to let the user simply select the distribution type (shape), it's central tendency (location), and variance (spread) --that would cover a lot of use-cases.

**Comment by OAGr** [comment]
My guess is that we'll eventually want a more visual editor, where you can drag parts of the graph.

Some existing interactive normal distributions:
http://www.intmath.com/counting-probability/normal-distribution-graph-interactive.php
http://www.measuringu.com/normal_curve.php
https://academo.org/demos/gaussian-distribution/

**Comment by davidmanheim** [comment]
If you want to go in that direction, it's really an elicitation probelm - @bobcolner specifying variance is something people are not good at intuitively.Instead, you really would want an elicitation tool, like Match; http://optics.eee.nottingham.ac.uk/match/uncertainty.php

Without that, people are better at selecting whether something is logically bounded or not, and what the biggest and littlest value they can imagine occurring are (5% and 95% - but people will still underimagine the range, per most probability elicitation studies.)

This is why I think that sticking with the [a,b] syntax, and simply adding <x,y> as logical bounds makes more sense from a UI perspective for most users - and those sophisticated enough to want a specific distribution and to input something specific about it can, and perhaps should, be building their models in R or Python anyways.

**Comment by davidmanheim** [comment]
OK, here is a tentative lists of all the distribution types I think we need to support for normal users; (with a suggested distribution in parentheses), and some notes

- Normal (Normal); Already implemented
- Uniform (Uniform); Already Implemented
- Fat Tailed (Cauchy); Nice because finding the parameters is analytical, and simulation exists in jstat. For Cauchy, Location is the average of the two points, and the Scale is the difference between the two points over ~12.6
- Lower bounded (Lognormal, Gamma, or Exponential? Gamma is most general.); Exponential can take only one input. 
- Interval bounded (Beta)

Additional suggestions?

**Comment by OAGr** [comment]
- Cauchy seems very easy to implement (in UI), as it's symmetric and won't need many more parameters.  I imagine we'll need some explanations in how cauchy distributions are different from normal distributions (when are good cases to use each).

- For exponential distributions, maybe we use the 90% bound?  This would stick with the 90% interval (in this case between 0 and whatever number they say).

- What about Pareto distributions?

- Binary distributions of different kinds.  Some are already supported by the function library, but this is a good place for them well.  One issue I have with binary distributions is that many can be reduced to continuous distributions extra truncation.  So instead of thinking 'random integers from 1-5', I'd prefer 'a uniform distribution rounded to random integers, from 1-5'. 

- I like the idea of classifying the 'lower bounded' distributions in one group, at least the ones with multiple parameters.  Does anyone else have opinions on Lognormal vs. Gamma?

I think that for now the most important next distribution to add is either Lognormal or Gamma. They're so similar that either that honestly, for common use by most people, changing them seems be very unlikely to change a decision.

Which do you think is more popular, or representative of common things.  For instance, which would you expect to be a better fit for the following things:
- Minutes of a Youtube video
- Revenue of a business
- Start up expected earnings
- Expected time of commuting.

**Comment by OAGr** [comment]
Also, one way of making this more flexible now, is to focus on instead allowing the user to input raw data.  This would allow users to use existing tools to generate whatever curve or data they want.

**Comment by cecemel** [comment]
+1 for adding raw data. Plot it, let user (roughly) compare it with the different distributions available. (or even use statistical tests to see wether normal,  exponential, gamma if existent). Else, normal users (including me) could be scared off.

**Comment by davidmanheim** [comment]
I think the first thing to do is to replace stochator with jStat; there is almost nothing listed here that won't be relatively straightforward once the underlying package with different distributions is available.

It looks like this just requires replacing function sampleNormal from src/lib/engine/distribution.js
Once that is done, it's easy to template other distributions from that.

(The call for a single sample from the normal distribution is; jStat.normal.sample( mean, std )
The call for a single sample from a cauchy is; jStat.cauchy.sample( local, scale )
etc.)

**Comment by davidmanheim** [comment]
@cecemel - I'd guess that being able to toggle between them and see the graphs would work to see which is appropriate, no?

**Comment by OAGr** [comment]
One small point (probably):

I imagine we don't need sampleNormal to use jStat, in order to have other distributions use jStat.  How about we just begin by adding one long-tail distribution type in jStat, then if that works we refactor after?

**Comment by kuza55** [comment]
I would personally be interested in using this to do some basic risk modeling for my own interest, and I would be interested in seeing some distributions that work with 3-point estimates.

**Comment by yalla** [comment]
Is there some update on that topic? I'm currently building a model for estimating staff utilization (basic application of Little's Theorem) and to build a proper model I'd need the Poisson distribution (for customer arrival rate).

**Comment by OAGr** [comment]
We don't yet have native support, but we do now have custom data input, which allows any distribution you could get from other sources.

https://github.com/getguesstimate/guesstimate-app/pull/78

**Comment by yalla** [comment]
Good to know! I just helped myself with generating my desired Poisson samples in R. For the curious, here's how I did it:

```
poidata <- rpois(5000, lambda=5)
write.table(t(poidata), file="~/poidata.txt", sep=",", row.names = FALSE, col.names = FALSE)
```
Then `cat` on the commandline on the file, copy, paste, voila.

**Comment by OAGr** [comment]
Happy to hear that worked!  

I'd love to do this sooner, but feel like building it in the app will take some time to get right.  I want to think through the interface for entering distributions, then custom distributions.  Private models seemed more important (which goes along with privacy), then I would like to do a little bit to make the app more understandable to new users, then more distribution types.

**Comment by davidmanheim** [comment]
IS it simple enough to use the hack described above as a way to enter
distributions; treat all as custom distributions, but allow a syntax for
calling, say, probdist (see: https://www.npmjs.com/package/probdist  /
https://github.com/rodrigosetti/probdist) distributions with pass through
arguments, and have it generate samples that way.

**Comment by OAGr** [comment]
I guess it would be on the easy side to have something temporary for users who are ok with writing out the function name and don't want documentation.  If we could do this in a day or so, it seems like a decent option for now.

**Comment by yalla** [comment]
@OAGr without looking at the source, what does the resampling part of the custom data function do? Does it derive a "real" PDF or do you simply throw in a lot of random numbers, Monte Carlo style, and just pick those random samples under the curve until you have enough samples? If yes, how did it know that I only gave integers (Poisson) and no real numbers? I'm asking this because I naively created 5000 samples myself, which significantly slowed the app down. So would it be enough just throwing in enough samples so that the curve becomes visible?

**Comment by OAGr** [comment]
It's done by sampling, there's no fitting at this point.  So if you enter 5 custom data values, or 500, it will resample these 5,000 times.  These values will then be randomly sampled from for each next computation.  This is a bit redundant and I expect to improve this.

Can you link to the model where the speed is an issue?  I could imagine it could be slow to save (as each save saves the entire graph, which would include all those samples), but would expect the processing to be decent.

**Comment by yalla** [comment]
@OAGr The model in question is at http://www.getguesstimate.com/models/3206 - but it turned out that a restart of my webbrowser solved the problem. Never mind the speed issues. Sorry, my bad.

**Comment by OAGr** [comment]
This PR deals with most of the jStat distributions.  https://github.com/getguesstimate/guesstimate-app/pull/156

Also, we have custom data now.

**Comment by mmcdermott** [comment]
We also now support using the lognormal distribution straight from the confidence interval input method. Please let us know if there are any disribution use cases not covered by the added functions and the lognormal functionality.

**Comment by OAGr** [comment]
I'm pretty sure we support enough distributions by this point for future requests to be specific (I want this specific kind of distribution).
